pkg/meta/base.go

```diff
@@ -2218,6 +2218,66 @@ func (m *baseMeta) CompactAll(ctx Context, threads int, bar *utils.Bar) syscall.
 	return 0
 }
 
+func (m *baseMeta) Compact(ctx Context, inode Ino, concurrency int, preFunc, postFunc func()) syscall.Errno {
+	var attr Attr
+	if st := m.GetAttr(ctx, inode, &attr); st != 0 {
+		logger.Errorf("get attr error [inode %v]: %v", inode, st)
+		return st
+	}
+
+	var wg sync.WaitGroup
+	// compact: consumer
+	chunkChan := make(chan cchunk, 10000)
+	for i := 0; i < concurrency; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for c := range chunkChan {
+				m.en.compactChunk(c.inode, c.indx, true)
+				postFunc()
+			}
+		}()
+	}
+
+	type fileNode struct {
+		ino Ino
+		len uint64
+	}
+	fileChan := make(chan *fileNode, 10000)
+
+	// dispatch: producer
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for file := range fileChan {
+			// calc in local
+			chunkCnt := uint32((file.len + ChunkSize - 1) / ChunkSize)
+			for i := uint32(0); i < chunkCnt; i++ {
```

Comment:
do this in the walk() will be simpler

Reply:
Considering large files, this approach may be better. When the compact chunk is stuck, we can proactively scan the next batch of files using engine.doReadDir, and it doesn't require a very large chunk channel.

it can change to

``` go
// scan
st := m.walk(ctx, inode, "", &attr, func(ctx Context, fIno Ino, path string, fAttr *Attr) {
    if fAttr.Typ != TypeFile {
	    return
    }
    // calc chunk index in local
    chunkCnt := uint32((fAttr.Length + ChunkSize - 1) / ChunkSize)
    for i := uint32(0); i < chunkCnt; i++ {
	    preFunc()
	    chunkChan <- cchunk{inode: fIno, indx: i}
    }
})

// finish
close(chunkChan)
wg.Wait()
```